src/k2/httpproxy/HTTPProxy.cpp

```diff
@@ -347,10 +296,118 @@ HTTPProxy::_handleGetSchema(shd::GetSchemaRequest&& request) {
+    values.reserve(shExpr.valueChildren.size());
+    for (auto& val: shExpr.valueChildren) {
+        values.push_back(getValue(std::move(val)));
+    }
+    std::vector<dtoexp::Expression> exprs;
+    exprs.reserve(shExpr.expressionChildren.size());
+    for (auto& cexpr: shExpr.expressionChildren) {
+        exprs.push_back(getFilterExpression(std::move(cexpr)));
+    }
+    return dtoexp::makeExpression(static_cast<dtoexp::Operation>(to_integral(shExpr.op)), std::move(values), std::move(exprs));
+}
+
 seastar::future<std::tuple<sh::Status, shd::CreateQueryResponse>>
 HTTPProxy::_handleCreateQuery(shd::CreateQueryRequest&& request) {
     K2LOG_D(log::httpproxy, "Received create query request {}", request);
-    return MakeHTTPResponse<shd::CreateQueryResponse>(sh::Statuses::S501_Not_Implemented("create query not implemented"), shd::CreateQueryResponse{});
+    auto it = _txns.find(request.timestamp);
+    if (it == _txns.end()) {
+        return MakeHTTPResponse<shd::CreateQueryResponse>(Txn_S410_Gone, shd::CreateQueryResponse{});
+    }
+    return _client.createQuery(request.collectionName, request.schemaName)
+        .then([this, req=std::move(request)] (auto&& result) mutable {
+            if(!result.status.is2xxOK()) {
+                return MakeHTTPResponse<shd::CreateQueryResponse>(sh::Status{.code = result.status.code, .message = result.status.message}, shd::CreateQueryResponse{});
+            }
+            try {
+                setQueryRecord(req.collectionName, std::move(req.key), result.query.startScanRecord);
+                setQueryRecord(req.collectionName, std::move(req.endKey), result.query.endScanRecord);
+
+                if (req.recordLimit >= 0) result.query.setLimit(req.recordLimit);
+                if (req.includeVersionMismatch) result.query.setIncludeVersionMismatch(req.includeVersionMismatch);
+                if (req.reverseDirection) result.query.setReverseDirection(req.reverseDirection);
+                if (req.filterExpression.op != shdexp::Operation::UNKNOWN) {
+                    dtoexp::Expression expr = getFilterExpression(std::move(req.filterExpression));
+                    result.query.setFilterExpression(std::move(expr));
+                }
+                if (req.projection.size() > 0) {
+                    std::vector<String> projection;
+                    projection.reserve(req.projection.size());
+                    for (sh::String& p : req.projection) {
+                        projection.push_back(String(std::move(p)));
+                    }
+                    result.query.addProjection(projection);
+                }
+            } catch(shd::DeserializationError& err) {
+                return MakeHTTPResponse<shd::CreateQueryResponse>(sh::Statuses::S400_Bad_Request(err.what()), shd::CreateQueryResponse{});
+            }
+
+            auto queryId = _queryID++;
+            _txns.at(req.timestamp).queries[queryId] = std::move(result.query);
```

Comment:
this can throw.

Reply:
It cannot throw as previously in line 349 it checks if timestamp txn exists.

Reply:
Done